connections: Add hole punching vision document (#323)
libp2p/specs: connections/hole-punching.md

connection to B.

### Private Browser (A) to Public Non-Browser (B)

Given that B is publicly routable A can establish a direct connection, though
multiple restrictions are imposed by the browser platform type of A:

- The only transport protocol allowing direct connections on browser platforms
  is WebSockets. Thus B has to support the WebSocket transport protocol.

- Browser platforms do not allow insecure WebSocket connections, thus B has to
  have a valid TLS certificate to offer secure Websocket communication.

In cases where B does not fulfill both of these requirements, additional steps
as detailed in [#private-browser-a-to-private-non-browser-b] are necessary.

### Public or Private Non-Browser (A) to Private Non-Browser (B)

In order to establish a direct connection from A to B, one needs to utilize some
means of firewall / NAT hole punching. Given the permissiveness of non-browser
platforms there are many ways to achieve this goal, with the below being one of
them.

- B uses the [AutoNat] protocol to determine whether it is running behind
  firewalls and/or NATs and in the latter case what type of NATs.
  

>   Note: Hole punching fails if either A or B runs behind a [symetric
>   NAT][symetric-nat].

Reviewer: ```suggestion
  Note: Hole punching fails if either A or B runs behind a [symmetric
  NAT][symmetric-nat].
```

- B establishes a TCP or QUIC connections to one or more relay servers and
  listens for incoming connection requests via the [circuit relay v2
  protocol][circuit-relay-v2]. Once B established a reservation at one or more
  relay servers, B can advertise its relayed addresses (e.g.
  `/ip4/.../tcp/.../p2p/QmRelay`) via some external mechanism.

  Note on decentralization: Given the low resource requirements of the [circuit
  relay v2 protocol][circuit-relay-v2] any public host can serve as a relay
  server.

> - A discovers the relayed address of B through some external mechanism. Given

Reviewer: Maybe provide an example of an external mechanism?

Author: Good idea. I added a paragraph on the notion of a discovery mechanism in https://github.com/libp2p/specs/pull/323/commits/21a3a11463568170a4ccef684857be17b45f7578. Let me know what you think.

Author: Resolving here. Please feel free to un-resolve.

>   the relayed address A can establish a relayed connection to B via the [circuit

Reviewer: ```suggestion
  the relayed address, A can establish a relayed connection to B via the [circuit
```

  relay v2 protocol][circuit-relay-v2] protocol over TCP or QUIC. A and B can
  then run the [Direct Connection Upgrade through Relay][DCUTR] protocol over
  the relayed connection to coordinate TCP or QUIC hole punching in the best
  case establishing a direction connection.

For more details see the [Project Flare][project-flare] project proposal.

### Private Browser (A) to Private Non-Browser (B)

As mentioned above browser platforms (1) do not allow libp2p implementations to
manage the network sockets nor (2) do they allow managing the transport security
stack, thus not allowing _insecure_ connections.

- Both A and B discover [STUN] servers through some external mechanism. Given
  that two hosts do not need to share the same [STUN] server to establish a
  WebRTC connection, this does not require any coordination beyond the [STUN]
  server discovery process.

- B establishes a TCP or QUIC connections to one or more relay servers and
  listens for incoming connection requests via the [circuit relay v2

>   protocol][circuit-relay-v2]. Once B established a reservation at one or more

Reviewer: ```suggestion
  protocol][circuit-relay-v2]. Once B establishes a reservation at one or more
```

  relay servers, B can advertise its relayed addresses (e.g.
  `/ip4/.../tcp/.../p2p/QmRelay`) via some external mechanism.

  Note: Given that browser platforms do not allow insecure connections, the

>   relay server needs listen on TLS secured Websockets.

Author: ```suggestion
  relay server needs to listen on TLS secured Websockets.
```

  B's relayed address A can establish a Websocket connection to the relay server
  of B. With the help of the [circuit relay v2 protocol][circuit-relay-v2]
  protocol A and B can exchange [SDP] messages eventually allowing A B and to
  establish a WebRTC connection.

## FAQ

- **Why not connect to a relay server via WebRTC instead of Websockets?**

  WebRTC needs some mechanism to exchange [SDP] messages. Unless there is some
  other means to exchange these [SDP] messages to a relay server, one has to use
  Websockets.

- **What to do when hole punching fails?**

  Peers will be able to communicate via the established [circuit relay v2
  protocol][circuit-relay-v2] connection. Though this connection is limited in
  both transfered bytes and time. For now, upper layer protocols need to cope
  with the fact that full connectivity can not be guaranteed.

- **Why use both [AutoNAT] and [STUN], why not settle on one for both TCP / QUIC
  and WebRTC based hole punching?**
  
  On browser platforms libp2p implementations do not control the local WebRTC
  stack. This same stack can only operate with [STUN] and not [AutoNAT]. One

>   could use [STUN] instead of [AutoNAT] for TCP and QUIC hole punching though.

Reviewer: So why do we have AutoNAT instead of STUN?  I clicked the link, but see it's a TODO to write this spec.

Author: I don't think I am the right person to reply here as I have not been involved in the design nor in one of the implementations. If I recall correctly one consideration was to prevent non-libp2p free-riding. @vyzo can you add more details here?

Reviewer 2: in part yes, but it was more motivated for having pervasive support with no dedicated server.
Any public node can provide the AutoNAT service.

Note that AutoNAT is used only for detecting public reachability; the observed address inference from identify complements it.

[TURN]: https://en.wikipedia.org/wiki/Traversal_Using_Relays_around_NAT
[STUN]: https://en.wikipedia.org/wiki/STUN
[AutoNAT]: https://github.com/libp2p/specs/issues/180
[SDP]: https://en.wikipedia.org/wiki/Session_Description_Protocol
[circuit-relay-v1]: https://github.com/libp2p/specs/tree/master/relay
[circuit-relay-v2]: https://github.com/libp2p/specs/issues/314
[DCUTR]: https://github.com/libp2p/specs/pull/173
[project-flare]: https://github.com/protocol/web3-dev-team/pull/21

> [symetric-nat]: https://dh2i.com/kbs/kbs-2961448-understanding-different-nat-types-and-hole-punching/

Reviewer: ```suggestion
[symmetric-nat]: https://dh2i.com/kbs/kbs-2961448-understanding-different-nat-types-and-hole-punching/
```

# Hole punching

| Lifecycle Stage | Maturity      | Status | Latest Revision |
|-----------------|---------------|--------|-----------------|
| 1A              | Working Draft | Active | r0, 2021-04-27  |

Authors: [@mxinden]

Interest Group:

> [@mxinden]: https://github.com/mxinden

Author: Great! Added in a4330f46dcb8a42ed5cb7fd2212418b868ec08d5

See the [lifecycle document][lifecycle-spec] for context about maturity level
and spec status.

[lifecycle-spec]: https://github.com/libp2p/specs/blob/master/00-framework-01-spec-lifecycle.md

## Overview

This document describes the process of **establishing direct connections between
two libp2p peers in which one or both are behind firewalls and/or NATs**. The
document derives many concepts from the _Interactive Connectivity Establishment
(ICE) standard. The most notable deviation from ICE is the goal of hole punching
in a decentralized fashion.

Additional goals:
- Interoperability across platforms and implementations.
- Shared abstractions reusable across transport protocols.

This document does not reflect the status quo, but much rather the high level
long term vision. To gain an overview of the status quo, see [issue #312].

[issue #312]: https://github.com/libp2p/specs/issues/312

## Definitions

> ### Routability

Author: :+1: Renamed in 43592cfbc98d52dfef6d181a35c0e139f8f66ed3.

### Private Non-Browser (A) to Public Non-Browser (B)

Given that B is publicly routable, A can establish a direct TCP or QUIC
connection to B.

### Private Browser (A) to Public Non-Browser (B)

Given that B is publicly routable A can establish a direct connection, though
multiple restrictions are imposed by the browser platform type of A:

- The only transport protocol allowing direct connections on browser platforms
  is WebSockets. Thus B has to support the WebSocket transport protocol.

- Browser platforms do not allow insecure WebSocket connections, thus B has to
  have a valid TLS certificate to offer secure Websocket communication.

In cases where B does not fulfill both of these requirements, additional steps
as detailed in [#private-browser-a-to-private-non-browser-b] are necessary.

### Public or Private Non-Browser (A) to Private Non-Browser (B)

In order to establish a direct connection from A to B, one needs to utilize some
means of firewall / NAT hole punching. Given the permissiveness of non-browser
platforms there are many ways to achieve this goal, with the below being one of
them.

- B uses the [AutoNat] protocol to determine whether it is running behind

>   firewalls and/or NATs and in the latter case what type of NATs.

Author: Thanks for catching this @vyzo. Would you mind helping me rephrase this paragraph to make sure it is accurate? Unfortunately I am not familiar with AutoNAT.

  
  Note: Hole punching fails if either A or B runs behind a [symmetric
  NAT][symmetric-nat].

- B establishes a TCP or QUIC connections to one or more relay servers and
  listens for incoming connection requests via the [circuit relay v2
  protocol][circuit-relay-v2]. Once B established a reservation at one or more
  relay servers, B can advertise its relayed addresses (e.g.
  `/ip4/.../tcp/.../p2p/QmRelay`) via some external mechanism.

  Note on decentralization: Given the low resource requirements of the [circuit
  relay v2 protocol][circuit-relay-v2] any public host can serve as a relay
  server.

- A discovers the relayed address of B through some external mechanism. Given
  the relayed address, A can establish a relayed connection to B via the [circuit
  relay v2 protocol][circuit-relay-v2] protocol over TCP or QUIC. A and B can
  then run the [Direct Connection Upgrade through Relay][DCUTR] protocol over
  the relayed connection to coordinate TCP or QUIC hole punching in the best
  case establishing a direction connection.

For more details see the [Project Flare][project-flare] project proposal.

> ### Private Browser (A) to Private Non-Browser (B)

Author: Good point. I extended `# Reachability` in db48e616df459d256e72f569a860ce341869483f. Let me know what you think.

> ### Private Browser (A) to Private Non-Browser (B)

Author: Resolving here. Please feel free to un-resolve.

This document does not reflect the status quo, but much rather the high level
long term vision. To gain an overview of the status quo, see [issue #312].

[issue #312]: https://github.com/libp2p/specs/issues/312

## Definitions

### Reachability

This document distinguishes between two types of hosts: publicly reachable and
non-publicly reachable hosts. Addresses of the former type of hosts can be
directly dialed on the public Internet. Addresses of the latter type of hosts
can only be dialed through a relay protocol, or through some means of hole
punching. The document refers to publicly reachable hosts as _public_ and
non-publicly reachable hosts as _private_.

Note: (For now) this document ignores the case where two hosts are not publicly
reachable, but operate within the same local layer 2 or layer 3 network.
Oftentimes in such cases direct connections between the two hosts can be
established without prior hole punching.

### Platforms

We differentiate two types of platforms: _browser_ and _non-browser_.

On _non-browser_ platforms a libp2p implementation has extensive control over
the (TCP or UDP) network socket, e.g. controlling port-reuse. The libp2p

> implementation controls the full transport security stack. Further more, unless

Reviewer 3: ```suggestion
implementation controls the full transport security stack. Furthermore, unless
```

run behind a Firewall or NAT, _non-browser_ platform hosts can be directly
dialed through their publicly reachable addresses.

> On the contrary, _browser_ platforms a libp2p implementation does not manage the

Reviewer 3: ```suggestion
On the contrary, on _browser_ platforms a libp2p implementation does not manage the
```

namely QUIC and WebRTC.

#### Relay

A relay protocol allows two clients to communicate, routing data via a relay
server, instead of using a direct connection. Examples of relay protocols are
[TURN], [circuit relay v1][circuit-relay-v1] and [circuit relay
v2][circuit-relay-v2].

#### STUN-like

STUN-like protocols allow a host to detect whether they are running behind a
firewall and/or NAT and if so they enable a host to discover their perceived
public IP addresses as well as their port mapping. Examples of STUN-like
protocols are [STUN] (hence the name) and [AutoNAT].

#### Coordination

Coordination protocols run on top of some relay protocol enabling two hosts to
coordinate the hole punching process. Examples of coordination protocols are the
[Session Description Protocol (SDP)][SDP] and [Direct Connection Upgrade through
Relay][DCUTR].

### Discovery mechanism

A discovery mechanism allows a node to discover peers in the network. A node can
use such mechanism e.g. to find peers at random, to learn external addresses of
peers, or more targeted to discover peers that provide a specific service. Among
others one can build a discovery mechanism on top of the libp2p [Kademlia] and

> [Gossipsub] protocol. The [IPFS's usage of a Kademlia DHT][ipfs-kademlia] is a

Reviewer 3: ```suggestion
[Gossipsub] protocol. [IPFS's usage of a Kademlia DHT][ipfs-kademlia] is a
```

platform type (browser / non-browser) as as well as routability (public /
private). The below describes on a high level how to establish direct
connections across the various permutations of the two dimensions.

### Public Non-Browser (A) to Public Non-Browser (B)

This is the ideal case, where both hosts can be directly dialed, thus there
being no need for hole punching. The two hosts should use either TCP or QUIC to
communicate directly.

### Private Non-Browser (A) to Public Non-Browser (B)

Given that B is publicly reachable, A can establish a direct TCP or QUIC
connection to B.

### Private Browser (A) to Public Non-Browser (B)

Given that B is publicly reachable A can establish a direct connection, though
multiple restrictions are imposed by the browser platform type of A:

- The only transport protocol allowing direct connections on browser platforms
  is WebSockets. Thus B has to support the WebSocket transport protocol.

- Browser platforms do not allow insecure WebSocket connections, thus B has to
  have a valid TLS certificate to offer secure Websocket communication.

In cases where B does not fulfill both of these requirements, additional steps

> as detailed in [#private-browser-a-to-private-non-browser-b] are necessary.

Reviewer 3: The link doesn't work.

firewall and/or NAT and if so they enable a host to discover their perceived
public IP addresses as well as their port mapping. Examples of STUN-like
protocols are [STUN] (hence the name) and [AutoNAT] in combination with
[Identify].

#### Coordination

Coordination protocols run on top of some relay protocol enabling two hosts to
coordinate the hole punching process. Examples of coordination protocols are the
[Session Description Protocol (SDP)][SDP] and [Direct Connection Upgrade through
Relay][DCUTR].

### Discovery mechanism

A discovery mechanism allows a node to discover peers in the network. A node can
use such mechanism e.g. to find peers at random, to learn external addresses of
peers, or more targeted to discover peers that provide a specific service. Among
others one can build a discovery mechanism on top of the libp2p [Kademlia] and
[Gossipsub] protocol. [IPFS's usage of a Kademlia DHT][ipfs-kademlia] is a
good example of a discovery mechanism.

Further details on discovery mechanisms are out of scope for this document. The
descriptions below will assume a generic _external discovery mechanism_ to be
available.

## Vision

> When it comes to establishing direct connections one needs to differentiate by
> platform type (browser / non-browser) as as well as routability (public /
> private). The below describes on a high level how to establish direct
> connections across the various permutations of the two dimensions.

Reviewer 4: Perhaps we could replace `Non-Browser` with `Native`?

```suggestion
When it comes to establishing direct connections one needs to differentiate by
runtime type (JS in browser / native code) as as well as routability (public /
private). The below describes on a high level how to establish direct
connections across the various permutations of the two dimensions.
```

Author: Still agree with the standpoint that `Non-Browser` is not ideal, though I am not sure `Native` is a good replacement. To me `Native` is something _native_ to a specific OS, Architecture or Processor similar to the definition on Wikipedia:

> In computing, native software or data-formats are those that were designed to run on a particular operating system. In a more technical sense, native code is code written specifically for a certain processor.[1] In contrast, cross-platform software can be run on multiple operating systems and/or computer architectures. 

https://en.wikipedia.org/wiki/Native_(computing)

This would not hold for libp2p NodeJS given that OS or Processor specific properties are abstracted via the V8 virtual machine.

Does that reasoning make sense @lidel?

  listens for incoming connection requests via the [circuit relay v2
  protocol][circuit-relay-v2]. Once B established a reservation at one or more
  relay servers, B can advertise its relayed addresses (e.g.
  `/ip4/.../tcp/.../p2p/QmRelay`) via some external mechanism.

  Note on decentralization: Given the low resource requirements of the [circuit
  relay v2 protocol][circuit-relay-v2] any public host can serve as a relay
  server.

- A discovers the relayed address of B through some external mechanism. Given
  the relayed address, A can establish a relayed connection to B via the [circuit
  relay v2 protocol][circuit-relay-v2] protocol over TCP or QUIC. A and B can
  then run the [Direct Connection Upgrade through Relay][DCUTR] protocol over
  the relayed connection to coordinate TCP or QUIC hole punching in the best
  case establishing a direction connection.

For more details see the [Project Flare][project-flare] project proposal.

### Private Browser (A) to Private Non-Browser (B)

As mentioned above browser platforms (1) do not allow libp2p implementations to
manage the network sockets nor (2) do they allow managing the transport security

> stack, thus not allowing _insecure_ connections.

Reviewer 4: I  find sending this doc useful when people want to understand what this "secure/insecure" division in browsers is about:

```suggestion
stack, thus not allowing _insecure_ connections from [secure contexts](https://developer.mozilla.org/en-US/docs/Web/Security/Secure_Contexts).
```

Author: :+1: good idea. Included via https://github.com/libp2p/specs/pull/323/commits/7b494d6684a0edf3adae51080066ab15d3733797.
